File: frontend/src/features/gestion-comision/types/analisis.ts
/**
 * types/analisis.ts — TypeScript types mirroring backend Pydantic schemas
 * for analisis endpoints (C-11).
 *
 * Source of truth: backend/app/schemas/analisis.py
 * No `any` allowed.
 */

// ---------------------------------------------------------------------------
// Atrasados — GET /api/v1/analisis/atrasados
// ---------------------------------------------------------------------------

export interface AlumnoAtrasado {
  alumno_id: string
  nombre: string
  apellidos: string
  actividades_faltantes: string[]
  actividades_reprobadas: string[]
}

export interface AtrasadosResponse {
  atrasados: AlumnoAtrasado[]
  sin_padron: boolean
}

// ---------------------------------------------------------------------------
// Ranking — GET /api/v1/analisis/ranking
// ---------------------------------------------------------------------------

export interface RankingItem {
  alumno_id: string
  nombre: string
  apellidos: string
  aprobadas: number
}

export interface RankingResponse {
  items: RankingItem[]
}

// ---------------------------------------------------------------------------
// Notas finales — GET /api/v1/analisis/notas-finales
// ---------------------------------------------------------------------------

export interface NotaFinalItem {
  alumno_id: string
  nombre: string
  apellidos: string
  aprobadas: number
  total_actividades: number
  porcentaje_aprobacion: number
}

export interface NotasFinalesResponse {
  items: NotaFinalItem[]
}

// ---------------------------------------------------------------------------
// Reporte — GET /api/v1/analisis/reporte
// ---------------------------------------------------------------------------

export interface ReporteAsignacion {
  total_alumnos: number
  total_atrasados: number
  pct_aprobacion_general: number
  total_actividades: number
  tiene_datos: boolean
}

File: backend/app/schemas/analisis.py
"""
app/schemas/analisis.py — Pydantic v2 schemas for analisis endpoints.

All schemas use ConfigDict(extra='forbid') as per project rules.

Implemented: C-11 (analisis-atrasados-reportes)
Updated:     C-23 (monitor-general) — MonitorGeneralItem, MonitorGeneralResponse
"""
from __future__ import annotations

import uuid

from pydantic import BaseModel, ConfigDict


class _StrictBase(BaseModel):
    model_config = ConfigDict(extra="forbid")


# ---------------------------------------------------------------------------
# 1. Atrasados
# ---------------------------------------------------------------------------

class AlumnoAtrasado(_StrictBase):
    alumno_id: uuid.UUID
    nombre: str
    apellidos: str
    actividades_faltantes: list[str]
    actividades_reprobadas: list[str]


class AtrasadosResponse(_StrictBase):
    atrasados: list[AlumnoAtrasado]
    sin_padron: bool


# ---------------------------------------------------------------------------
# 2. Ranking
# ---------------------------------------------------------------------------

class RankingItem(_StrictBase):
    alumno_id: uuid.UUID
    nombre: str
    apellidos: str
    aprobadas: int


class RankingResponse(_StrictBase):
    items: list[RankingItem]


# ---------------------------------------------------------------------------
# 3. Notas finales
# ---------------------------------------------------------------------------

class NotaFinalItem(_StrictBase):
    alumno_id: uuid.UUID
    nombre: str
    apellidos: str
    aprobadas: int
    total_actividades: int
    porcentaje_aprobacion: float


class NotasFinalesResponse(_StrictBase):
    items: list[NotaFinalItem]


# ---------------------------------------------------------------------------
# 4. Reporte
# ---------------------------------------------------------------------------

class ReporteAsignacion(_StrictBase):
    total_alumnos: int
    total_atrasados: int
    pct_aprobacion_general: float
    total_actividades: int
    tiene_datos: bool


# ---------------------------------------------------------------------------
# 5. Monitor general (F2.7) — C-23
# ---------------------------------------------------------------------------

class MonitorGeneralItem(_StrictBase):
    alumno_id: uuid.UUID
    nombre: str
    apellidos: str
    email: str
    materia: str
    comision: str | None
    cohorte: str
    actividades_aprobadas: int
    actividades_totales: int
    estado: str  # "al_dia" | "atrasado" | "sin_datos"


class MonitorGeneralResponse(_StrictBase):
    items: list[MonitorGeneralItem]
    total: int
